![](https://storage.googleapis.com/mvps1_24/logo01.jpg)

![](https://storage.googleapis.com/mvps1_24/line01_02.jpg)

## Introdução

Este trabalho consiste na construção de um pipeline de dados utilizando tecnologias na nuvem. O pipeline irá envolver a busca e **coleta** de dados armazenando-os na nuvem, a construção de um **modelo de dados** juntamente com um **Catálogo de Dados** contendo minimamente uma descrição detalhada dos dados e seus domínios, a **carga** dos dados para um Data Lake utilizando pipelines de ETL (Extração, Transformação e Carga) nesta plataforma Databricks, **análise** dos dados e no final uma **autoanálise**.

## Objetivo
O objetivo é ver quais atores e atrizes mais participaram de filmes premiados com Oscar, relacionando eles às produções cinematográficas às quais atuaram. Neste banco de dados as produções estão entre os anos 1960 até 2024. Um segundo objetivo seria comparar a pontuação IMDB, que é dada pelo público, com o número de oscars que um filme ganhou.

## Descrição do problema
Serão usadas originalmente duas bases de dados da Internet Movie Database (IMDb), cujos atributos são mostrados nas tabelas abaixo e no detalhamento. A primeira base de dados contém títulos de filmes produzidos entre os anos 1960 e 2024 com seus respectivos códigos. A segunda base de dados contém nomes de artistas com os respectivos códigos dos filmes em que atuaram. 

![](https://storage.googleapis.com/mvps1_24/tab_basedados2.jpg)

Irei trabalhar estas duas tabelas para alinha-las a este projeto, detalhado em seguida.

**Ao término deste projeto iremos poder responder aos seguintes questionamentos:**

- Quais foram os atores / atrizes que mais participaram de produções vencedoras de Oscars.
- Comparar a relação existente entre indicações e premiações ao Oscar de atores e atrizes atuantes.
- O comparar a pontuação IMDb e numero de oscars de filmes.














![](https://storage.googleapis.com/mvps1_24/line01_03b.jpg)


O IMDb é um banco de dados, é uma fonte confiável de informações sobre cinema e televisão e é o maior banco de dados de cinema do mundo. Tem cadastro gratuito de avaliações de filmes, séries, jogos e muitos outros conteúdos.

As classificações do IMDb são médias ponderadas das avaliações dos usuários, e não médias de dados brutos. Isso significa que nem todas as classificações têm o mesmo impacto na classificação final, pois depende do número de votos para cada filme. A pontuação do IMDb é baseada na avaliação de usuários registrados, que podem dar notas de 1 a 10 para filmes, séries e outros conteúdos. Essas notas são agregadas e resumidas em uma única classificação.

As _nominations_, são o número de indicações na IMDb e incluem indicações para diferentes prêmios, em categorias variadas, evidenciando o sucesso e a repercussão da produção, bem como as indicações para concorrer ao Oscar. 
As indicações do IMDb são baseadas nas avaliações dos usuários do site, e também na atividade deles na lista de observação, enquanto que a escolha dos indicados ao Oscar é feita por membros da Academia de Artes e Ciências Cinematográficas.

Curiosidade: um critério usado é que uma nota 8 ou superior significa que o filme é bom ou tem alta probabilidade de ser bom, uma nota entre 8 e 7 as coisas podem variar muito e abaixo de 7 o filme não é muito bom.


![](https://storage.googleapis.com/mvps1_24/linha.jpg)

### Detalhamento

Coletei as duas bases de dados do site da Internet Movie Database, sendo que uma delas carreguei no Google Cloud Storage.

https://datasets.imdbws.com/name.basics.tsv.gz
https://storage.googleapis.com/mvps1_24/world_imdb_movies_top_movies_per_year_t4.csv.gz

A primeira base de dados contém todos os títulos de filmes produzidos entre os anos 1960 e 2024 com seus respectivos códigos, ano de lançamento, duração, pontuação IMDb, número de votos, indicações ao Oscar e número de Oscars que o filme ganhou. A segunda base de dados contém um número bem grande de nomes de atores, o código do filme em que eles atuaram, o ano de nascimento e falecimento, a profissão principal deles. Nem todos estes atores participaram de alguma produção entre os anos 1960 e 2024. Uma parte deles está associada a um ou mais códigos de filmes em que atuaram naquele período.


####Projeto Conceitual
O projeto conceitual tem o seguinte diagrama abaixo. A primeira tabela contém nomes de atores com seus respectivos códigos e tb os códigos dos filmes em que atuaram. A segunda tabela contém titulo dos filmes entre 1960 e 2024 com seus respectivos códigos. Os códigos dos filmes nas duas tabelas são iguais.

![](https://storage.googleapis.com/mvps1_24/prj_conceitual3.jpg)

Em um filme pode ter atuado no mínimo um ator ou no máximo muitos atores. Por outro lado um ator pode ter atuado no minimo em um filme ou no máximo em vários filmes. Isto configura um **relacionamento de muitos para muitos (N:N)** e neste caso a regra nos manda criar uma terceira tabela relacionando as duas primeiras tabelas, resultando como abaixo:

- **Tabela Filme**: com os atributos [PK]**cod_filme** e titulo.
- **Tabela Ator**: com um identificador único (por exemplo, [PK]**cod_ator**) e o nome.
- **Tabela Atua** (Relacionamento): que contém os campos [FK]**cod_filme** e [FK]**cod_ator** para representar o relacionamento muitos-para-muitos entre filmes e atores. 

Esta separação evita a redundância, pois cada ator é cadastrado uma única vez, e o relacionamento é feito através de chaves estrangeiras.

####Projeto Lógico
Neste meu projeto porém, irei considerar que as tabelas não serão alteradas no futuro e que somente o relacionamento do ator com o filme é o que me interessa, isto me permite simplificar o modelo lógico usando apenas duas tabelas:

- **Tabela Filmes**: com **cod_filme**, título, ano, duração, pts_imdb, n_votos, indicações, n_oscars.
- **Tabela Atores**: preparada para ser a tabela de relacionamento, com dois atributos: **cod_filme**, referenciando o filme, e **nome_ator**.

![](https://storage.googleapis.com/mvps1_24/prj_logico3.jpg)

Neste modelo, o nome do ator aparece quantas vezes forem necessárias, para cada filme no qual ele atuou. 
Embora isso já represente a relação muitos-para-muitos, a desvantagem é que o nome dos atores pode ficar 
repetido em diversas linhas. Esse design funciona bem se realmente não formos alterar ou expandir as tabelas. Irei usar a técnica de JOIN, que seria a junção entre as duas tabelas e poder responder.

####Descrição dos dados, seus domínios, valores mínimos e máximos esperados.
Baseado nos atributos das base de dados, segue a tabela com os tipos de dados e valores:   

![](https://storage.googleapis.com/mvps1_24/tab_basedados3.jpg)

Todos os campo não aceitam valor nulo, devem ser NOT NULL, menos o campo duração, ele permite valor nulo.
![](https://storage.googleapis.com/mvps1_24/linha.jpg)

####Programação

In [0]:
import io               #Library fornece as ferramentas para leitura e escrita de arquivos.
import requests         #Library utilizada para fazer requisições HTTP.
import gzip             #Para a descompactação de arquivos.

from pyspark.sql import SparkSession    # para poder baixar o arquivo compactado
from pyspark.sql.functions import col   # para poder ordenar as colunas

No quadro 11 abaixo criei um dataframe para conter o arquivo com os dados dos atores. O arquivo está em um repositório da Internet Movie Database (IMDb) e é do tipo CSV e está compactado em **tsv.gz**. Foi necessário também especificar a codificação tipo UTF-8.

Todo o processo de **carga** e descompactação está documentado abaixo e é feito com Pyspark.

In [0]:
# Criar uma instância de SparkSession de nome spark
spark = SparkSession.builder.appName("LerArquivoGzipDeURL").getOrCreate()
# SparkSession.builder: é o construtor
# getOrCreate(): cria uma nova SparkSession ou retorna uma existente, caso já tenha sido criada previamente.

# URL do arquivo Gzip
url = "https://datasets.imdbws.com/name.basics.tsv.gz"

# Fazer o download do arquivo gzip usando requests
response = requests.get(url)

# Descompactar o arquivo no formato Gzip
with gzip.GzipFile(fileobj=io.BytesIO(response.content)) as filedescomp:
    # Carregar os dados descompactados em um DataFrame do Spark
    # gzip.GzipFile: classe usada para ler ou escrever arquivos compactados no formato Gzip.
    # fileobj=io.BytesIO(response.content): cria um objeto de arquivo em memória, que contém os dados HTTP (response.content)

    data = filedescomp.read().decode("utf-8")
    rdd = spark.sparkContext.parallelize(data.splitlines())
    # O método parallelize pega o texto e distribui entre os nós do cluster do Spark para serem processados paralelamente.
    # O método splitlines() divide uma string em várias linhas (quebras de linhas \n) onde cada linha é um elemento.
    ator_spark_df = spark.read.csv(rdd, sep="\t", header=True)

    ator_spark_df.printSchema()

    #NOME DO DATAFRAME:     ator_spark_df


root
 |-- nconst: string (nullable = true)
 |-- primaryName: string (nullable = true)
 |-- birthYear: string (nullable = true)
 |-- deathYear: string (nullable = true)
 |-- primaryProfession: string (nullable = true)
 |-- knownForTitles: string (nullable = true)



Análise deste dataframe: **ator_spark_df**

In [0]:
ator_spark_df = ator_spark_df.orderBy(col("nconst").asc())    

# Exibindo o conteúdo do DataFrame Spark na ordem ascendente
ator_spark_df.show(20)

+---------+-------------------+---------+---------+--------------------+--------------------+
|   nconst|        primaryName|birthYear|deathYear|   primaryProfession|      knownForTitles|
+---------+-------------------+---------+---------+--------------------+--------------------+
|nm0000001|       Fred Astaire|     1899|     1987|actor,miscellaneo...|tt0072308,tt00504...|
|nm0000002|      Lauren Bacall|     1924|     2014|actress,soundtrac...|tt0037382,tt00752...|
|nm0000003|    Brigitte Bardot|     1934|       \N|actress,music_dep...|tt0057345,tt00491...|
|nm0000004|       John Belushi|     1949|     1982|actor,writer,musi...|tt0072562,tt00779...|
|nm0000005|     Ingmar Bergman|     1918|     2007|writer,director,a...|tt0050986,tt00694...|
|nm0000006|     Ingrid Bergman|     1915|     1982|actress,producer,...|tt0034583,tt00381...|
|nm0000007|    Humphrey Bogart|     1899|     1957|actor,producer,mi...|tt0034583,tt00432...|
|nm0000008|      Marlon Brando|     1924|     2004|actor,dir

Visualizando as 1000 primeiras linhas dos dados acima, ordenados pelo código do ator (nconst) em modo ascendente, aparentemente estão limpos. No próximo passo vou inverter para a ordem descendente, para visualizar a outra ponta dos dados. 

In [0]:
ator_spark_df = ator_spark_df.orderBy(col("nconst").desc())    

# Exibindo o conteúdo do DataFrame Spark na ordem descendente
ator_spark_df.show(50)

+---------+--------------------+---------+---------+--------------------+--------------------+
|   nconst|         primaryName|birthYear|deathYear|   primaryProfession|      knownForTitles|
+---------+--------------------+---------+---------+--------------------+--------------------+
|nm9993719|          Andre Hill|       \N|       \N|                  \N|                  \N|
|nm9993718|         Aayush Nair|       \N|       \N|     cinematographer|           tt8736744|
|nm9993717|  Harikrishnan Rajan|       \N|       \N|     cinematographer|           tt8736744|
|nm9993716|       Essias Loberg|       \N|       \N|                  \N|                  \N|
|nm9993714|   Romeo del Rosario|       \N|       \N|animation_departm...|tt11657662,tt1406...|
|nm9993713|       Sambit Mishra|       \N|       \N|     writer,producer|tt20319332,tt2784...|
|nm9993712|     Corny O'Connell|       \N|       \N|                  \N|                  \N|
|nm9993711|       David Gluzman|       \N|       \

####Qualidade dos dados deste primeiro set.
**Análise (ator_spark_df).**

**-** Neste primeiro dataset notam-se alguns defeitos. Pela coluna _primaryName_, que é o nome do ator, notamos que tem atores/atrizes que não estão associados a nenhum título de filme. Estas tuplas precisam ser eliminadas. Outro defeito é que para vários atores/atrizes não consta o ano de nascimento. Estas tuplas também precisam ser eliminadas.

**-** As duas colunas, _primaryProfession_ e _knownForTitles_ são multivaloradas. As colunas _primaryProfession_ e _deathYear_ não interessam para este projeto, então irei eliminá-las também. A coluna _knownForTitles_ contém os códigos dos filmes e irei separa-los colocando cada um deles em uma célula, de modo que o nome do ator estará associado às produções em que atuou e o nome dele se repetirá pelo número de produções em que participou.

**-** A coluna _nconst_ é o código de cada ator/atriz. Esta coluna também não interessa para este projeto e irei elimina-la.

No próximo passo farei o desmembramento da coluna multivalorada _knownForTitles_ para fazer um relacionamento ator/filmes.

In [0]:
# Na tabela com nomes dos atores, temos dados multivalorados, ou seja, em uma tupla com a célula contendo o nome 
# de um ator a coluna correspondente com o código das produções, contém vários códigos. Irei separar esses códigos
# para ficarem um em cada célula. 
from pyspark.sql.functions import explode
from pyspark.sql.functions import split

# similar a: SQL SELECT nconst, primaryName, knownForTitles... - só que o conteúdo da coluna knownForTitles é separado por virgulas.
atores_rel_filmes_spark_df = ator_spark_df.select(ator_spark_df.nconst, ator_spark_df.primaryName, ator_spark_df.birthYear, explode(split(ator_spark_df.knownForTitles, ',')).alias("id_tt"))

atores_rel_filmes_spark_df.show(50)      

#NOME DO DATAFRAME:     atores_rel_filmes_spark_df

+---------+------------------+---------+----------+
|   nconst|       primaryName|birthYear|     id_tt|
+---------+------------------+---------+----------+
|nm9993719|        Andre Hill|       \N|        \N|
|nm9993718|       Aayush Nair|       \N| tt8736744|
|nm9993717|Harikrishnan Rajan|       \N| tt8736744|
|nm9993716|     Essias Loberg|       \N|        \N|
|nm9993714| Romeo del Rosario|       \N|tt11657662|
|nm9993714| Romeo del Rosario|       \N|tt14069590|
|nm9993714| Romeo del Rosario|       \N| tt2455546|
|nm9993713|     Sambit Mishra|       \N|tt20319332|
|nm9993713|     Sambit Mishra|       \N|tt27843798|
|nm9993713|     Sambit Mishra|       \N|tt27191658|
|nm9993713|     Sambit Mishra|       \N|tt10709066|
|nm9993712|   Corny O'Connell|       \N|        \N|
|nm9993711|     David Gluzman|       \N|        \N|
|nm9993710| Nestor Rudnytskyy|       \N|        \N|
|nm9993709|         Lu Bevins|       \N|tt17717854|
|nm9993709|         Lu Bevins|       \N|tt11772904|
|nm9993709| 

Com a tabela acima nos aproximamos do objetivo. Na coluna _primaryName_ tem o nome do ator e a coluna _knownForTitles_ foi desmembrada e criei outra coluna _id_tt_ com os o código dos filmes separados um a um. Agora cada código de filme está associado a um ator. 

In [0]:
atores_rel_filmes_spark_df.printSchema()

root
 |-- nconst: string (nullable = true)
 |-- primaryName: string (nullable = true)
 |-- birthYear: string (nullable = true)
 |-- id_tt: string (nullable = false)



Vou criar agora outro dataframe para fazer a carga do segundo arquivo, contendo o nome das produções cinematográficas.

In [0]:
# URL do arquivo Gzip
url = "https://storage.googleapis.com/mvps1_24/world_imdb_movies_top_movies_per_year_t4.csv.gz"

# Fazer o download do arquivo gzip usando requests
response = requests.get(url)

# Descompactar o arquivo no formato Gzip
with gzip.GzipFile(fileobj=io.BytesIO(response.content)) as filedescomp:
    # Carregar os dados descompactados em um DataFrame do Spark
    # gzip.GzipFile: classe usada para ler ou escrever arquivos compactados no formato Gzip.
    # fileobj=io.BytesIO(response.content): cria um objeto de arquivo em memória, que contém os dados HTTP (response.content)

    data = filedescomp.read().decode("utf-8")
    rdd = spark.sparkContext.parallelize(data.splitlines())
    # O método parallelize pega o texto e distribui entre os nós do cluster do Spark para serem processados paralelamente.
    # O método splitlines() divide uma string em várias linhas (quebras de linhas \n) onde cada linha é um elemento.
    filmes_spark_df = spark.read.csv(rdd, sep=";", header=True, inferSchema=True)

    filmes_spark_df = filmes_spark_df.orderBy(col("year").desc()) 

    # Exibindo o conteúdo do DataFrame Spark na ordem ascendente
    filmes_spark_df.show(50)

    #NOME DO DATAFRAME:     filmes_spark_df

+----------+--------------------+----+--------+-----------+------+----------+-----+
|        id|               title|year|duration|rating_imdb|  vote|nomination|oscar|
+----------+--------------------+----+--------+-----------+------+----------+-----+
| tt0327785|   The Killer's Game|2024|  1h 44m|         57| 93000|         0|    0|
|tt12930350|My Spy The Eterna...|2024|  1h 52m|         57| 10000|         0|    0|
|tt10094388|         The Windigo|2024|  1h 25m|         53|   576|         0|    0|
|tt10128846|         Megalopolis|2024|  2h 18m|         48| 30000|         0|    0|
|tt10171472|            Goodrich|2024|  1h 50m|         63| 28000|         0|    0|
|tt10236164|The Girl with the...|2024|   2h 3m|         75|   967|        16|    0|
|tt10243672|                 Joy|2024|  1h 55m|         71| 66000|         0|    0|
|tt10245072|         Salem's Lot|2024|  1h 54m|         56| 30000|         0|    0|
|tt10284944|             The Fix|2024|  1h 38m|         45|   855|         0

In [0]:
filmes_spark_df = filmes_spark_df.orderBy(col("year").asc()) 

# Exibindo o conteúdo do DataFrame Spark na ordem ascendente
filmes_spark_df.show(50)

+---------+--------------------+----+--------+-----------+-----+----------+-----+
|       id|               title|year|duration|rating_imdb| vote|nomination|oscar|
+---------+--------------------+----+--------+-----------+-----+----------+-----+
|tt0049360|         El impostor|1960|  1h 19m|         66|   28|         0|    0|
|tt0050971|            Sigfrido|1960|  1h 37m|         57|   82|         0|    0|
|tt0051536|     Devil's Partner|1960|  1h 14m|         52|  615|         0|    0|
|tt0051863|    Life Is a Circus|1960|  1h 24m|         78|  201|         0|    0|
|tt0051874|          La Llorona|1960|  1h 15m|         58|  157|         0|    0|
|tt0052555|The Vampire and t...|1960|  1h 25m|         56|  682|         0|    0|
|tt0052607|The Battle of the...|1960|  1h 24m|         66|18000|         0|    0|
|tt0052614|          Le bel �ge|1960|  1h 40m|         70|   78|         0|    0|
|tt0052635|          Mein Kampf|1960|  1h 51m|         76|  396|         0|    0|
|tt0052674|     

####Qualidade dos dados deste segundo set.
**Análise (filmes_spark_df).**

**-** Listei o dataframe acima com mais de 1000 linhas, tanto na ordem ascendente como na descendente e verifiquei que os dados estão tratados e não é preciso fazer ajustes.

In [0]:
filmes_spark_df.printSchema()

root
 |-- id: string (nullable = true)
 |-- title: string (nullable = true)
 |-- year: integer (nullable = true)
 |-- duration: string (nullable = true)
 |-- rating_imdb: integer (nullable = true)
 |-- vote: integer (nullable = true)
 |-- nomination: integer (nullable = true)
 |-- oscar: integer (nullable = true)



CRIAÇÃO DO DATABASE **BRONZE** E CARGA DE DADOS

In [0]:
%sql CREATE DATABASE IF NOT EXISTS Bronze;   -- Se não existir, será criado o DataBase Bronze, que irá guardar os dados "crus"

In [0]:
# Salvando o DataFrame actors_basics... no DataBase Bronze
atores_rel_filmes_spark_df.write.mode("overwrite").saveAsTable("Bronze.atores_rel_filmes")

In [0]:
# Salvando o DataFrame movies_info... no DataBase Bronze
filmes_spark_df.write.mode("overwrite").saveAsTable("Bronze.filmes")

In [0]:
%sql SELECT nconst AS cod_Ator, primaryName AS Nome_ator,birthYear AS Ano_nascimento, id_tt AS cod_Filme FROM Bronze.atores_rel_filmes ORDER BY nconst DESC LIMIT 60 

cod_Ator Nome_ator Ano_nascimento cod_Filme nm9993719 Andre Hill \N \N nm9993718 Aayush Nair \N tt8736744 nm9993717 Harikrishnan Rajan \N tt8736744 nm9993716 Essias Loberg \N \N nm9993714 Romeo del Rosario \N tt11657662 nm9993714 Romeo del Rosario \N tt14069590 nm9993714 Romeo del Rosario \N tt2455546 nm9993713 Sambit Mishra \N tt20319332 nm9993713 Sambit Mishra \N tt27843798 nm9993713 Sambit Mishra \N tt27191658 nm9993713 Sambit Mishra \N tt10709066 nm9993712 Corny O'Connell \N \N nm9993711 David Gluzman \N \N nm9993710 Nestor Rudnytskyy \N \N nm9993709 Lu Bevins \N tt17717854 nm9993709 Lu Bevins \N tt11772904 nm9993709 Lu Bevins \N tt11772812 nm9993709 Lu Bevins \N tt11697102 nm9993708 Eli Bevins \N tt36103776 nm9993708 Eli Bevins \N tt11702702 nm9993708 Eli Bevins \N tt14508814 nm9993708 Eli Bevins \N tt36042179 nm9993707 Alexandra Rago \N \N nm9993706 Michal Jackowski \N tt8744250 nm9993705 Bozena Zegert \N tt8744250 nm9993704 Wojciech Kolodziej \N tt8744250 nm9993703 James Craigmyle \N tt11212278 nm9993703 James Craigmyle \N tt6914160 nm9993703 James Craigmyle \N tt10627062 nm9993703 James Craigmyle \N tt6225166 nm9993702 Liz Matory \N tt8740182 nm9993701 Sanjai Kuriakose \N tt8736744 nm9993700 Sexy Angel \N tt7523066 nm9993699 Dani Jacob \N tt8736744 nm9993698 Sebi John \N tt8736744 nm9993697 Zakariya Ganim \N tt8744160 nm9993696 Ibrahim-Aloduley \N tt8744160 nm9993694 Chinmay Mishra \N tt18361688 nm9993694 Chinmay Mishra \N tt18687502 nm9993694 Chinmay Mishra \N tt8737752 nm9993694 Chinmay Mishra \N tt9083282 nm9993693 Apsara Rani 1996 tt12856788 nm9993693 Apsara Rani 1996 tt13847502 nm9993693 Apsara Rani 1996 tt8302382 nm9993693 Apsara Rani 1996 tt25385692 nm9993692 Shiela Martin \N tt7888884 nm9993691 Ursula Gehrmann \N tt7888884 nm9993690 David Jewell \N tt7888884 nm9993689 Chris Bailey \N tt31850564 nm9993688 Stefania Zadra \N tt7310898 nm9993687 Hiroko Ide \N tt7170282 nm9993686 Yukari Funaki \N tt7170282 nm9993685 John Root \N tt3741700 nm9993684 Angela Ng \N tt3741700 nm9993684 Angela Ng \N tt10182516 nm9993683 Lula Langston \N tt8744086 nm9993682 Cheryl Clark \N tt8744086 nm9993681 Saiya Palmer \N tt8295580 nm9993680 Christopher-Lawson Palmer \N tt21745568 nm9993680 Christopher-Lawson Palmer \N tt10427366

Ordenando a tabela acima pelo ano de nascimento do ator, vemos que tem artistas sem o ano de nascimento. Vemos também que tem atores que não estão associados a nenhum filme. Estas tuplas podem ser eliminandas retirando aquelas onde o _ano_ tenha outros caracteres ascii que não sejam números, e as produções que não contenham a letras 'tt' na sua composição, visto que o código de filmes começa sempre pelas letras 'tt'.

Outra observação é que a base de dados relaciona produções entre os anos 1960 e 2024, então vou considerar apenas as produções com atores nascidos 100 anos antes do primeiro premio, ou seja, entre 1860 e 2024.

In [0]:
%sql 
-- Retirando as tuplas onde o ano de nascimento contenham outros carateres que não sejam números.

SELECT nconst AS cod_Ator, primaryName AS Nome_ator, birthYear AS Ano_nascimento, id_tt AS cod_Filme FROM Bronze.atores_rel_filmes WHERE birthYear BETWEEN CHAR(48) AND CHAR(57);

cod_Ator Nome_ator Ano_nascimento cod_Filme nm2889095 Barbara Devil 1983 tt7050960 nm2889095 Barbara Devil 1983 tt1167409 nm2889095 Barbara Devil 1983 tt1316462 nm2889095 Barbara Devil 1983 tt1261878 nm2889076 Kazimierz Janecki 1925 tt1171264 nm2889076 Kazimierz Janecki 1925 tt0088188 nm2889076 Kazimierz Janecki 1925 tt0086224 nm2889076 Kazimierz Janecki 1925 tt21982996 nm2889060 Eerie Von 1964 tt0425141 nm2889060 Eerie Von 1964 tt1067090 nm2889051 Owl Chapman 1950 tt28349644 nm2889051 Owl Chapman 1950 tt28297406 nm2889051 Owl Chapman 1950 tt28692137 nm2889051 Owl Chapman 1950 tt28289217 nm2889045 David Fox 1977 tt12593682 nm2889045 David Fox 1977 tt1477834 nm2889045 David Fox 1977 tt2239822 nm2889045 David Fox 1977 tt1431045 nm2889021 Ann Pinnington 1938 tt1166817 nm2889011 Patrick Allen Cole 1981 tt0062595 nm2889011 Patrick Allen Cole 1981 tt0056758 nm2889009 Matt Oliver 1972 tt0481510 nm2889009 Matt Oliver 1972 tt1659198 nm2889009 Matt Oliver 1972 tt1838704 nm2889009 Matt Oliver 1972 tt5183652 nm2888990 Matt Reilly 1980 tt0430357 nm2888990 Matt Reilly 1980 tt0448157 nm2888990 Matt Reilly 1980 tt1055369 nm2888990 Matt Reilly 1980 tt7128316 nm2888980 Teresa Kossak 1934 tt1171264 nm2888980 Teresa Kossak 1934 tt2439690 nm2888980 Teresa Kossak 1934 tt0088188 nm2888980 Teresa Kossak 1934 tt1171260 nm2888958 Helena Postránecká 1947 tt0070452 nm2888958 Helena Postránecká 1947 tt0278628 nm2888958 Helena Postránecká 1947 tt0174790 nm2888958 Helena Postránecká 1947 tt0174811 nm2888950 Vijay Vinayak Joshi 1984 tt2882328 nm2888950 Vijay Vinayak Joshi 1984 tt3001638 nm2888950 Vijay Vinayak Joshi 1984 tt1860215 nm2888950 Vijay Vinayak Joshi 1984 tt1532567 nm2888941 Rochelle Rose 1989 tt1121790 nm2888897 Vladik 1990 \N nm2888878 Adam Palenta 1976 tt3753664 nm2888878 Adam Palenta 1976 tt3476308 nm2888878 Adam Palenta 1976 tt2147594 nm2888878 Adam Palenta 1976 tt5688178 nm2888846 Miroslav Milanovic Miki 1983 tt0332452 nm2888846 Miroslav Milanovic Miki 1983 tt1186830 nm2888846 Miroslav Milanovic Miki 1983 tt0408306 nm2888846 Miroslav Milanovic Miki 1983 tt0808339 nm2888839 Matt Turney 1925 tt4370946 nm2888839 Matt Turney 1925 tt1069951 nm2888839 Matt Turney 1925 tt0478971 nm2888839 Matt Turney 1925 tt0268824 nm2888838 Loek Hollander 1938 tt0070430 nm2888800 Krystyna Lason 1940 tt0076922 nm2888800 Krystyna Lason 1940 tt0259308 nm2888800 Krystyna Lason 1940 tt1187013 nm2888800 Krystyna Lason 1940 tt1187007 nm2888795 Ricardo Blei 1989 tt1112587 nm2888795 Ricardo Blei 1989 tt4145054 nm2888795 Ricardo Blei 1989 tt4944852 nm2888795 Ricardo Blei 1989 tt1119662 nm2888752 Nik Isbelle 1987 tt2226754 nm2888752 Nik Isbelle 1987 tt0412142 nm2888752 Nik Isbelle 1987 tt4200600 nm2888752 Nik Isbelle 1987 tt2197832 nm2888730 Eddie Hardy 1969 tt0117786 nm2888730 Eddie Hardy 1969 tt1288589 nm2888730 Eddie Hardy 1969 tt0112182 nm2888730 Eddie Hardy 1969 tt0402193 nm2888720 Theodore Mann 1924 tt0092322 nm2888720 Theodore Mann 1924 tt0271280 nm2888720 Theodore Mann 1924 tt0817467 nm2888711 Monika Munoz 1978 tt1172065 nm2888711 Monika Munoz 1978 tt6439386 nm2888711 Monika Munoz 1978 tt6337290 nm2888711 Monika Munoz 1978 tt0337742 nm2888643 Aina Jonsson 1934 tt1172058 nm2888643 Aina Jonsson 1934 tt2919866 nm2888624 Phil Jamieson 1977 tt0210584 nm2888624 Phil Jamieson 1977 tt0280605 nm2888624 Phil Jamieson 1977 tt0307750 nm2888624 Phil Jamieson 1977 tt35557844 nm2888576 Joy Pankin 1946 tt0088634 nm2888576 Joy Pankin 1946 tt0082814 nm2888576 Joy Pankin 1946 tt0085065 nm2888562 Bridget A. Schmidt 1966 \N nm2888524 Darrell Britt-Gibson 1985 tt5027774 nm2888524 Darrell Britt-Gibson 1985 tt9784798 nm2888524 Darrell Britt-Gibson 1985 tt4139124 nm2888524 Darrell Britt-Gibson 1985 tt4385888 nm2888474 Utako Nakayama 1893 tt4452096 nm2888474 Utako Nakayama 1893 tt4489908 nm2888474 Utako Nakayama 1893 tt4641878 nm2888474 Utako Nakayama 1893 tt4721578 nm2888472 Alejandro Hurtado Terán 1936 tt10397386 nm2888472 Alejandro Hurtado Terán 1936 tt10592186 nm2888472 Alejandro Hurtado T

In [0]:
%sql 
-- Retirada da tuplas em que o código do filme não começa com as letras 'tt'. 
-- Retirada das tuplas onde o ano de nascimento esteja fora do intervalo de 1860 e 2024 

SELECT nconst AS cod_Ator, primaryName AS Nome_ator, birthYear AS Ano_nascimento, id_tt AS cod_Filme FROM Bronze.atores_rel_filmes WHERE id_tt LIKE '%tt%' AND birthYear BETWEEN 1860 AND 2024 
GROUP BY cod_ator, primaryName, birthYear, id_tt
ORDER BY primaryName DESC
LIMIT 50

cod_Ator Nome_ator Ano_nascimento cod_Filme nm2513158 ​Rebekah Brooks 1968 tt0199257 nm2513158 ​Rebekah Brooks 1968 tt0981226 nm2513158 ​Rebekah Brooks 1968 tt0181260 nm2513158 ​Rebekah Brooks 1968 tt0983197 nm1188541 Þórðarson Garðar Bachmann 1986 tt0306763 nm1930089 Þórunn Magnea Magnúsdóttir 1945 tt9100822 nm1930089 Þórunn Magnea Magnúsdóttir 1945 tt0805576 nm1930089 Þórunn Magnea Magnúsdóttir 1945 tt8309886 nm1930089 Þórunn Magnea Magnúsdóttir 1945 tt1384929 nm1800527 Þórunn Lárusdóttir 1973 tt31862843 nm1800527 Þórunn Lárusdóttir 1973 tt9174804 nm1800527 Þórunn Lárusdóttir 1973 tt1245694 nm1800527 Þórunn Lárusdóttir 1973 tt15468716 nm4684035 Þórunn Guðlaugsdóttir 1986 tt0816692 nm4684035 Þórunn Guðlaugsdóttir 1986 tt26218798 nm4684035 Þórunn Guðlaugsdóttir 1986 tt0944947 nm4684035 Þórunn Guðlaugsdóttir 1986 tt2061851 nm0905090 Þórir Waagfjörð 1974 tt0114851 nm1802427 Þórhildur Þorleifsdóttir 1945 tt34389151 nm1802427 Þórhildur Þorleifsdóttir 1945 tt9420566 nm1802427 Þórhildur Þorleifsdóttir 1945 tt0369000 nm1802427 Þórhildur Þorleifsdóttir 1945 tt15468716 nm0841230 Þórhallur Sverrisson 1970 tt0399454 nm0841230 Þórhallur Sverrisson 1970 tt0840046 nm0841230 Þórhallur Sverrisson 1970 tt1327136 nm0841230 Þórhallur Sverrisson 1970 tt0260594 nm0480486 Þórhallur Sigurðsson 1947 tt0359950 nm0480486 Þórhallur Sigurðsson 1947 tt1930425 nm0480486 Þórhallur Sigurðsson 1947 tt0477474 nm0480486 Þórhallur Sigurðsson 1947 tt0086917 nm0294915 Þóra Friðriksdóttir 1933 tt0108176 nm0294915 Þóra Friðriksdóttir 1933 tt0086917 nm0294915 Þóra Friðriksdóttir 1933 tt0095172 nm0294915 Þóra Friðriksdóttir 1933 tt0131553 nm1096391 Þóra Dungal 1976 tt0132889 nm0096464 Þóra Borg 1907 tt0086917 nm0096464 Þóra Borg 1907 tt0179224 nm0096464 Þóra Borg 1907 tt0447672 nm0096464 Þóra Borg 1907 tt0350653 nm1804024 Þór Eldon 1962 tt0439436 nm1804024 Þór Eldon 1962 tt0461958 nm1804024 Þór Eldon 1962 tt14751462 nm1804024 Þór Eldon 1962 tt0109028 nm0897518 Þrúður Vilhjálmsdóttir 1973 tt1764275 nm0897518 Þrúður Vilhjálmsdóttir 1973 tt9100822 nm0897518 Þrúður Vilhjálmsdóttir 1973 tt3561180 nm0897518 Þrúður Vilhjálmsdóttir 1973 tt0237993 nm0348276 Þröstur Leó Gunnarsson 1961 tt0351461 nm0348276 Þröstur Leó Gunnarsson 1961 tt26218798 nm0348276 Þröstur Leó Gunnarsson 1961 tt15399970

In [0]:
filmes_spark_df.printSchema()

root
 |-- id: string (nullable = true)
 |-- title: string (nullable = true)
 |-- year: integer (nullable = true)
 |-- duration: string (nullable = true)
 |-- rating_imdb: integer (nullable = true)
 |-- vote: integer (nullable = true)
 |-- nomination: integer (nullable = true)
 |-- oscar: integer (nullable = true)



In [0]:
%sql 
-- retirada das tuplas onde o código de filme está NULL
SELECT id AS cod_Filme, title AS Titulo_filme, year AS Ano_criacao, duration AS Duracao, rating_imdb AS Pontos_IMDb, vote AS num_Votos, nomination AS Indicacoes, oscar AS n_Oscars FROM Bronze.filmes 
WHERE duration IS NOT NULL 
ORDER BY title ASC
LIMIT 50

cod_Filme Titulo_filme Ano_criacao Duracao Pontos_IMDb num_Votos Indicacoes n_Oscars tt1699720 !Women Art Revolution 2010 1h 23m 69 264 0 0 tt10620868 #Alive 2020 1h 38m 63 50000 0 0 tt0068152 $ 1971 2h 1m 63 3000 0 0 tt0060697 $1,000 on the Black 1966 1h 42m 59 418 0 0 tt0061313 $10,000 Blood Money 1967 1h 37m 61 749 0 0 tt0790799 $9.99 2008 1h 18m 67 35000 4 0 tt2614684 '71 2014 1h 39m 72 62000 31 2 tt7518786 '83 2021 2h 42m 75 42000 52 0 tt4417402 '85: The Greatest Team in Football History 2016 1h 30m 68 165 0 0 tt0065600 'Ctyri vrazdy stac�, drahousku' 1971 1h 43m 74 835 0 0 tt0067003 'Doc' 1971 1h 36m 62 21000 0 0 tt0074080 'Gator Bait 1973 1h 28m 54 15000 0 0 tt0102006 'Harry Lund' l�gger n�san i bl�t! 1991 1h 45m 45 144 0 0 tt0087710 'Men Olsenbanden var ikke d�d!' 1984 1h 29m 61 872 0 0 tt0217978 'R Xmas 2001 1h 25m 57 17000 3 0 tt0090557 'Round Midnight 1986 2h 13m 74 6000 13 0 tt0073697 'Sheba, Baby' 1975 1h 30m 56 19000 0 0 tt0118523 'Til There Was You 1997 1h 53m 48 3000 0 0 tt0069678 'Tis Pity She's a Whore 1971 1h 31m 63 504 0 0 tt0090556 'night, Mother 1986 1h 40m 75 26000 0 0 tt0349047 (T)Raumschiff Surprise - Periode 1 2004 1h 27m 55 17000 2 0 tt1132193 (Untitled) 2009 1h 36m 62 24000 0 0 tt0179184 **** 1967 25h 62 108 0 0 tt0092494 *batteries not included 1987 1h 47m 66 39000 0 0 tt0200855 ....mehri to ploio 1966 1h 29m 71 218 0 0 tt0081964 ...All the Marbles 1981 1h 53m 64 25000 0 0 tt0079515 ...And Give Us Our Daily Sex 1979 1h 24m 42 340 0 0 tt0068420 ...And Hope to Die 1972 1h 39m 67 899 0 0 tt0053739 ...And Suddenly It's Murder! 1960 1h 46m 70 405 0 0 tt1479352 ...Sabik kasalanan ba? 1986 2h 72 99 0 0 tt0056985 ...denn die Musik und die Liebe in Tirol 1963 1h 34m 41 34 0 0 tt0071078 ...verdammt, ich bin erwachsen 1974 1h 34m 65 34 0 0 tt0058866 008: Operation Exterminate 1965 1h 24m 51 179 0 0 tt4917224 09/nov 2017 1h 30m 45 48000 0 0 tt0053556 091 Polic�a al habla 1960 1h 36m 60 129 0 0 tt2198094 1+1=11 2012 1h 6m 0 null 0 0 tt0056795 1, 000 Shapes of a Female 1963 1h 19m 49 72 0 0 tt0066731 1,000 Convicts and a Woman 1971 1h 32m 41 292 0 0 tt2353767 1,000 Times Good Night 2013 1h 57m 70 83000 14 0 tt0078721 10 1979 2h 2m 61 20000 0 2 tt0201236 10 Brothers of Shaolin 1977 1h 30m 53 159 0 0 tt1179933 10 Cloverfield Lane 2016 1h 43m 72 366000 48 0 tt10251538 10 Days with Dad 2020 1h 44m 54 13000 0 0 tt0499603 10 Items or Less 2006 1h 22m 65 17000 2 0 tt0819354 10 Questions for the Dalai Lama 2006 1h 25m 74 767 0 0 tt0066730 10 Rillington Place 1971 1h 51m 75 10000 0 1 tt0147800 10 Things I Hate About You 1999 1h 37m 73 407000 13 0 tt7939428 10 Things We Should Do Before We Break Up 2020 1h 14m 47 16000 0 0 tt1715873 10 Years 2011 1h 40m 61 28000 0 0 tt0085121 10 to Midnight 1983 1h 41m 63 98000 0 0

Nas linhas 36 e 37 abaixo vou aplicar os mesmos filtros das consultas SQL feitas acima ao dataframe e em seguida atualizar o database. Vou também retirar a coluna de código do ator (_cod_Ator_).

In [0]:
# Retirada da tuplas em que o código do filme não começa com as letras 'tt'. 

atores_rel_filmes_spark_df = atores_rel_filmes_spark_df.filter((atores_rel_filmes_spark_df.id_tt.like('%tt%')))

In [0]:
# Retirada das tuplas onde o ano de nascimento esteja fora do intervalo de 1860 e 2024 

atores_rel_filmes_spark_df = atores_rel_filmes_spark_df.filter((atores_rel_filmes_spark_df["birthYear"] >= "1860") & (atores_rel_filmes_spark_df["birthYear"] <= "2024")).orderBy(col("primaryName").asc()).drop("nconst")

atores_rel_filmes_spark_df.show(50)


+--------------------+---------+----------+
|         primaryName|birthYear|     id_tt|
+--------------------+---------+----------+
|"Stone" Robert A ...|     1982| tt1648179|
|"Stone" Robert A ...|     1982| tt1872194|
|"Stone" Robert A ...|     1982| tt1980929|
|"Stone" Robert A ...|     1982| tt2392137|
|'Amarillo Slim' P...|     1928| tt0071269|
|'Amarillo Slim' P...|     1928| tt1483037|
|'Amarillo Slim' P...|     1928| tt1446673|
|'Amarillo Slim' P...|     1928|tt20871262|
|'Atlas' Ramachandran|     1942| tt1311658|
|'Atlas' Ramachandran|     1942| tt3433162|
|'Atlas' Ramachandran|     1942| tt0928194|
|'Atlas' Ramachandran|     1942| tt2333760|
|'Baby' Carmen De Rue|     1908| tt0004635|
|'Baby' Carmen De Rue|     1908| tt0007787|
|'Baby' Carmen De Rue|     1908| tt0008144|
|'Baby' Carmen De Rue|     1908| tt0003698|
|  'Baby' Paul Cullen|     1962| tt0275309|
|'Beatle' Bob Matonis|     1953| tt1327590|
|'Beatle' Bob Matonis|     1953| tt4741930|
|'Beatle' Bob Matonis|     1953|

No proximo passo irei eliminar a coluna do ano de nascimento dos atores.Neste momento não nos interessa mais, pois já retirei os atores que nasceram fora da faixa de 1860 a 2024, e a tabela irá se alinhar mais ao projeto.

In [0]:
atores_rel_filmes_spark_df = atores_rel_filmes_spark_df.drop("birthYear")

In [0]:
atores_rel_filmes_spark_df.printSchema()

root
 |-- primaryName: string (nullable = true)
 |-- id_tt: string (nullable = false)



CRIAÇÃO DO DATABASE **SILVER** E CARGA DE DADOS

In [0]:
%sql CREATE DATABASE IF NOT EXISTS Silver;  -- Cria o DataBase Silver apenas se não existir

In [0]:
atores_rel_filmes_spark_df.write.mode("overwrite").saveAsTable("Silver.atores_rel_filmes")
filmes_spark_df.write.mode("overwrite").saveAsTable("Silver.filmes")

In [0]:
%sql
SELECT COUNT(*)
FROM Silver.filmes

count(1) 33241

In [0]:
%sql SELECT primaryName AS Nome_ator ,id_tt AS cod_Filme FROM Silver.atores_rel_filmes
    ORDER BY primaryName ASC LIMIT 50

Nome_ator cod_Filme "Stone" Robert A Stone III tt1648179 "Stone" Robert A Stone III tt1872194 "Stone" Robert A Stone III tt1980929 "Stone" Robert A Stone III tt2392137 'Amarillo Slim' Preston tt0071269 'Amarillo Slim' Preston tt1483037 'Amarillo Slim' Preston tt1446673 'Amarillo Slim' Preston tt20871262 'Atlas' Ramachandran tt3433162 'Atlas' Ramachandran tt0928194 'Atlas' Ramachandran tt2333760 'Atlas' Ramachandran tt1311658 'Baby' Carmen De Rue tt0004635 'Baby' Carmen De Rue tt0007787 'Baby' Carmen De Rue tt0008144 'Baby' Carmen De Rue tt0003698 'Baby' Paul Cullen tt0275309 'Beatle' Bob Matonis tt4741930 'Beatle' Bob Matonis tt0472240 'Beatle' Bob Matonis tt9009510 'Beatle' Bob Matonis tt1327590 'Big Bengt' Erlandsson tt0466348 'Big Luther' Kent tt2374967 'Big Walter' Price tt0072225 'Big' LeRoy Mobley tt0305357 'Big' LeRoy Mobley tt1591493 'Big' LeRoy Mobley tt1693107 'Big' LeRoy Mobley tt1957867 'Big' Lee tt0069761 'Blue' Gene Tyranny tt0077414 'Blue' Gene Tyranny tt0085871 'Blue' Gene Tyranny tt0181332 'Blue' Gene Tyranny tt0203426 'Boley' Dancewicz tt0040904 'Boots' Hansen tt0063060 'Boots' Hansen tt0199248 'Brother' Frank Jares tt0046601 'Brother' Frank Jares tt0044283 'Brother' Frank Jares tt0044229 'Brother' Frank Jares tt0044969 'Bud' Schelling tt0045826 'Butch' Bucholtz tt0066913 'By' Dunham tt0057707 'By' Dunham tt0057298 'By' Dunham tt0059298 'By' Dunham tt0058626 'Chico' Hernandez tt0120746 'Chico' Hernandez tt0236493 'Chico' Hernandez tt0386140 'Chico' Hernandez tt0129332

Irei criar o database Gold e copiar as tabelas do DB Silver para o DB Gold, para finalmente fazer os questionamentos propostos pelo projeto.

CRIAÇÃO DO DATABASE **GOLD** E CARGA DE DADOS

In [0]:
%sql CREATE DATABASE IF NOT EXISTS Gold;  -- Cria o DataBase Gold apenas se não existir

In [0]:
filmes_spark_df.write.mode("overwrite").saveAsTable("Gold.filmes");
atores_rel_filmes_spark_df.write.mode("overwrite").saveAsTable("Gold.atores_rel_filmes");

In [0]:
atores_rel_filmes_spark_df.printSchema()

root
 |-- primaryName: string (nullable = true)
 |-- id_tt: string (nullable = false)



In [0]:
filmes_spark_df.printSchema()

root
 |-- id: string (nullable = true)
 |-- title: string (nullable = true)
 |-- year: integer (nullable = true)
 |-- duration: string (nullable = true)
 |-- rating_imdb: integer (nullable = true)
 |-- vote: integer (nullable = true)
 |-- nomination: integer (nullable = true)
 |-- oscar: integer (nullable = true)



Verificamos acima que os atributos das duas tabelas estão de acordo com o projeto.

![](https://storage.googleapis.com/mvps1_24/linha.jpg)

####Solução do problema

**Respostas aos questionamentos propostos.**
![](https://storage.googleapis.com/mvps1_24/linha.jpg)

Na linha baixo vou fazer a primeira consulta: fornescendo alguns nomes de atores e atrizes para ver quem ganhou mais oscar e pontos.

**1)** Quais foram os atores e atrizes que mais participaram de produções que ganharam Oscars?

Vou escolher os seguintes atores: Jack Nicholson, Tom Cruise, Kevin Bacon, Meryl Streep, Anthony Hopkins, Emma Stone, Kate Winslet e Nicole Kidman.

In [0]:
%sql 
SELECT primaryName AS Nome_ator, title AS Titulo_filme, year AS Ano, rating_imdb AS Pontos_IMDb, vote AS n_Votos, nomination AS Indicacoes, oscar AS n_Oscars FROM Gold.atores_rel_filmes 
INNER JOIN Gold.filmes ON Gold.atores_rel_filmes.id_tt = Gold.filmes.id 
WHERE primaryName IN ("Jack Nicholson", "Tom Cruise", "Kevin Bacon", "Meryl Streep", "Anthony Hopkins", "Emma Stone", "Kate Winslet", "Nicole Kidman")
ORDER BY primaryName 
LIMIT 50

Nome_ator Titulo_filme Ano Pontos_IMDb n_Votos Indicacoes n_Oscars Anthony Hopkins The Silence of the Lambs 1991 86 16000000 50 0 Anthony Hopkins The World's Fastest Indian 2005 78 61000 6 0 Anthony Hopkins Hitchcock 2012 68 80000 29 1 Anthony Hopkins Nixon 1995 71 33000 18 4 Emma Stone La La Land 2016 80 703000 307 0 Emma Stone Easy A 2010 70 427000 22 0 Emma Stone The Help 2011 81 506000 121 0 Emma Stone Poor Things 2023 78 320000 414 0 Jack Nicholson Chinatown 1974 81 360000 24 0 Jack Nicholson As Good as It Gets 1997 77 325000 54 0 Jack Nicholson The Departed 2006 85 15000000 141 0 Jack Nicholson One Flew Over the Cuckoo's Nest 1975 87 11000000 15 0 Kate Winslet Titanic 1997 79 13000000 83 0 Kate Winslet Eternal Sunshine of the Spotless Mind 2004 83 11000000 111 0 Kate Winslet Revolutionary Road 2008 73 229000 73 3 Kate Winslet The Reader 2008 76 266000 48 0 Kevin Bacon Footloose 1984 66 96000 6 2 Kevin Bacon Hollow Man 2000 59 141000 15 1 Kevin Bacon The Woodsman 2004 71 36000 20 0 Kevin Bacon Mystic River 2003 79 497000 144 0 Meryl Streep Out of Africa 1985 71 88000 30 0 Meryl Streep August: Osage County 2013 72 97000 66 2 Meryl Streep Sophie's Choice 1982 75 54000 14 0 Meryl Streep The Devil Wears Prada 2006 69 489000 53 2 Nicole Kidman Moulin Rouge! 2001 76 304000 133 0 Nicole Kidman The Hours 2002 75 143000 126 0 Nicole Kidman Rabbit Hole 2010 70 52000 42 1 Nicole Kidman Dogville 2003 80 162000 33 0 Tom Cruise Top Gun 1986 69 518000 9 0 Tom Cruise Minority Report 2002 76 596000 91 1 Tom Cruise Jerry Maguire 1996 73 294000 41 0 Tom Cruise Mission: Impossible 1996 72 482000 17 0

**Comentário**: Na consulta acima foi usado a técnica de JOIN, uma junção entre as duas tabelas. Ordenando pelo número de Oscar notamos que, entre estes artistas escolhidos, o ator Kevin Bacon participou mais vezes de produções premiadas com Oscar, ao passo que o filme Poor Things teve 414 indicações, mas não ganhou Oscars. Lembro que as _indicações_ incluem indicações para diferentes prêmios, em categorias variadas.

![](https://storage.googleapis.com/mvps1_24/linha.jpg)
**2)** Quais foram os atores / atrizes que mais participaram de produções com mais indicações ao Oscar? Lembrando que o intervalo de premiações está entre os anos 1960 e 2024.

In [0]:
%sql 
SELECT primaryName AS Nome_ator, title AS Titulo_filme, year AS Ano, duration AS Duracao, (rating_imdb / 10) AS Pontos_IMDb, vote AS n_Votos,nomination AS Indicacoes, oscar AS n_Oscars FROM Gold.atores_rel_filmes 
INNER JOIN Gold.filmes ON Gold.atores_rel_filmes.id_tt = Gold.filmes.id 
WHERE oscar > 0
ORDER BY oscar DESC
LIMIT 50


Nome_ator Titulo_filme Ano Duracao Pontos_IMDb n_Votos Indicacoes n_Oscars Carl Anderson The Color Purple 1985 2h 34m 7.7 100000 25 11 Steve Ray The Color Purple 1985 2h 34m 7.7 100000 25 11 Paul Whiteman The Color Purple 1985 2h 34m 7.7 100000 25 11 Snooky Young The Color Purple 1985 2h 34m 7.7 100000 25 11 Marianna Tcherkassky The Turning Point 1977 1h 59m 6.8 53000 18 11 Speckled Red The Color Purple 1985 2h 34m 7.7 100000 25 11 Porter Grainger The Color Purple 1985 2h 34m 7.7 100000 25 11 Starr Danias The Turning Point 1977 1h 59m 6.8 53000 18 11 Adolph Caesar The Color Purple 1985 2h 34m 7.7 100000 25 11 Thomas A. Dorsey The Color Purple 1985 2h 34m 7.7 100000 25 11 Quincy Jones The Color Purple 1985 2h 34m 7.7 100000 25 11 Tom Bähler The Color Purple 1985 2h 34m 7.7 100000 25 11 Marshal Royal The Color Purple 1985 2h 34m 7.7 100000 25 11 Táta Vega The Color Purple 1985 2h 34m 7.7 100000 25 11 Rae Dawn Chong The Color Purple 1985 2h 34m 7.7 100000 25 11 Whoopi Goldberg The Color Purple 1985 2h 34m 7.7 100000 25 11 Jurgen Schneider The Turning Point 1977 1h 59m 6.8 53000 18 11 Willard E. Pugh The Color Purple 1985 2h 34m 7.7 100000 25 11 Rebecca Walker The Color Purple 1985 2h 34m 7.7 100000 25 11 Martine Van Hamel The Turning Point 1977 1h 59m 6.8 53000 18 11 Reuben Cannon The Color Purple 1985 2h 34m 7.7 100000 25 11 Aggie Guerard Rodgers The Color Purple 1985 2h 34m 7.7 100000 25 11 Richard Cragun The Turning Point 1977 1h 59m 6.8 53000 18 11 Maurice Moore The Color Purple 1985 2h 34m 7.7 100000 25 11 Richard Hazard The Color Purple 1985 2h 34m 7.7 100000 25 11 Henry Busse The Color Purple 1985 2h 34m 7.7 100000 25 11 Richard Pryor Jr. The Color Purple 1985 2h 34m 7.7 100000 25 11 Menno Meyjes The Color Purple 1985 2h 34m 7.7 100000 25 11 Rod Temperton The Color Purple 1985 2h 34m 7.7 100000 25 11 Akosua Busia The Color Purple 1985 2h 34m 7.7 100000 25 11 Roy Gaines The Color Purple 1985 2h 34m 7.7 100000 25 11 Mikhail Baryshnikov The Turning Point 1977 1h 59m 6.8 53000 18 11 Saax Bradbury The Turning Point 1977 1h 59m 6.8 53000 18 11 Ken Chase The Color Purple 1985 2h 34m 7.7 100000 25 11 Sam Nestico The Color Purple 1985 2h 34m 7.7 100000 25 11 Mikhail Fokin The Turning Point 1977 1h 59m 6.8 53000 18 11 Sandra Crouch The Color Purple 1985 2h 34m 7.7 100000 25 11 Alex Brown The Color Purple 1985 2h 34m 7.7 100000 25 11 Nananne Porcher The Turning Point 1977 1h 59m 6.8 53000 18 11 Enrique Martinez The Turning Point 1977 1h 59m 6.8 53000 18 11 Nathan Scott The Color Purple 1985 2h 34m 7.7 100000 25 11 Alexander Minz The Turning Point 1977 1h 59m 6.8 53000 18 11 Oprah Winfrey The Color Purple 1985 2h 34m 7.7 100000 25 11 Kenneth MacMillan The Turning Point 1977 1h 59m 6.8 53000 18 11 Alexandra Danilova The Turning Point 1977 1h 59m 6.8 53000 18 11 Henry Lange The Color Purple 1985 2h 34m 7.7 100000 25 11 Alice Walker The Color Purple 1985 2h 34m 7.7 100000 25 11 Leonard Jackson The Color Purple 1985 2h 34m 7.7 100000 25 11 Allen Daviau The Color Purple 1985 2h 34m 7.7 100000 25 11 Chris Boardman The Color Purple 1985 2h 34m 7.7 100000 25 11

Comentário: Temos acima a lista dos atores e atrizes (listei 5000) que mais participaram de filmes premiados, mas se observa outra coisa interessante. Pelo critério popular citado no início, os filmes com pontuação IMDb entre 7 e 8, são considerados neutros. Ordenando pela colna de número de Oscars nota-se que a grande maioria dos filmes que ganharam mais Oscar tem pontuação IMDb neutra. Isto pode ser devido por exemplo a altas expectativas sobre o filme. Filmes que ganham muitos Oscars geralmente vêm com uma grande expectativa. Pode ser também a diferença de público. O IMDb reflete a opinião de um público global e diversificado, enquanto os Oscars são decididos por um grupo específico de votantes da indústria cinematográfica.

![](https://storage.googleapis.com/mvps1_24/linha.jpg)
**3)** Comparar a pontuação IMDb e número de oscars de filmes. Qual a coerência entre a pontuação e a premiação do Oscar.

In [0]:
%sql 
SELECT title AS Titulo_filme, duration AS Duracao, (rating_imdb / 10) AS Pontos_IMDb, vote AS n_Votos, nomination AS Indicacoes, oscar AS n_Oscars FROM Gold.atores_rel_filmes 
INNER JOIN Gold.filmes ON Gold.atores_rel_filmes.id_tt = Gold.filmes.id  
WHERE rating_imdb > 0 AND oscar > 0
GROUP BY title, duration, rating_imdb, vote, nomination, oscar
ORDER BY rating_imdb DESC 
LIMIT 50

Titulo_filme Duracao Pontos_IMDb n_Votos Indicacoes n_Oscars The Shawshank Redemption 2h 22m 9.3 3000000 42 7 Fight Club 2h 19m 8.8 24000000 38 1 The Green Mile 3h 9m 8.6 15000000 37 4 Se7en 2h 7m 8.6 19000000 44 1 City of God 2h 10m 8.6 824000 50 4 American History X 1h 59m 8.5 12000000 15 1 The Prestige 2h 10m 8.5 15000000 44 2 For Sama 1h 40m 8.5 13000 52 1 Psycho 1h 49m 8.5 741000 14 4 Spider-Man: Across the Spider-Verse 2h 20m 8.5 428000 164 1 The Hole 2h 11m 8.5 21000 4 2 Das Boot 2h 29m 8.4 274000 12 6 Avengers: Infinity War 2h 29m 8.4 13000000 81 1 Woman in the Dunes 2h 27m 8.4 24000 3 2 Capernaum 2h 6m 8.4 113000 55 1 Avengers: Endgame 3h 1m 8.4 13000000 133 1 Memento 1h 53m 8.4 14000000 59 2 The Salt of the Earth 1h 50m 8.4 24000 15 1 Star Wars: Episode VI - Return of the Jedi 2h 11m 8.3 11000000 23 4 Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb 1h 35m 8.3 532000 11 4 Toy Story 1h 21m 8.3 11000000 24 3 Hoop Dreams 2h 50m 8.3 29000 9 1 The Hunt 1h 55m 8.3 380000 73 1 Am�lie 2h 2m 8.3 810000 74 5 Requiem for a Dream 1h 42m 8.3 925000 69 1 Incendies 2h 11m 8.3 218000 20 1 The Rescue 1h 47m 8.3 22000 39 1 Macario 1h 31m 8.3 46000 2 1 The Weavers: Wasn't That a Time 1h 18m 8.3 226 0 1 Scenes from a Marriage 2h 49m 8.3 92000 5 1 The Given Word 1h 38m 8.3 41000 0 1 The Look of Silence 1h 43m 8.3 14000 45 1 Promises 1h 46m 8.3 27000 5 1 The Sixth Sense 1h 47m 8.2 11000000 56 6 Hamlet 2h 20m 8.2 3000 6 2 Rocco and His Brothers 2h 59m 8.2 23000 10 2 The Great Escape 2h 52m 8.2 265000 12 1 The Elephant Man 2h 4m 8.2 266000 22 8 The Wolf of Wall Street 3h 8.2 16000000 179 5 A Clockwork Orange 2h 16m 8.2 901000 26 4 Yojimbo 1h 50m 8.2 136000 2 1 Casino 2h 58m 8.2 582000 11 1 Batman Begins 2h 20m 8.2 16000000 79 1 Kill Bill: Vol. 1 1h 51m 8.2 12000000 103 5 Spider-Man: No Way Home 2h 28m 8.2 926000 71 1 Prisoners 2h 33m 8.2 849000 38 1 Die Hard 2h 12m 8.2 977000 8 4 The Truman Show 1h 43m 8.2 12000000 69 3 Taxi Driver 1h 54m 8.2 958000 21 4 Full Metal Jacket 1h 56m 8.2 811000 15 1

Comentário: Na consulta acima confirmamos o comentário da consulta anterior. Outro motivo desta diferença pode ser devido aos temas e estilo. Muitos filmes premiados no Oscar têm temas artísticos, históricos ou dramáticos que podem não agradar a todos os públicos. Enquanto críticos e a Academia valorizam esses aspectos, o público geral pode achar o filme "difícil" ou "lento", resultando em avaliações mais neutras.


![](https://storage.googleapis.com/mvps1_24/linha.jpg)

In [0]:
%sql 
SELECT title AS Titulo_filme, duration AS Duracao, oscar AS n_Oscars, (rating_imdb / 10) AS Pontos_IMDB,
    CASE 
        WHEN rating_imdb > 80 THEN 'Bom'
        WHEN rating_imdb < 70 THEN 'Ruim'
        ELSE 'Neutro'
    END AS Classif
FROM Gold.filmes 
WHERE rating_imdb > 0 
ORDER BY oscar DESC 
LIMIT 50

Titulo_filme Duracao n_Oscars Pontos_IMDB Classif The Turning Point 1h 59m 11 6.8 Ruim The Color Purple 2h 34m 11 7.7 Neutro Gangs of New York 2h 47m 10 7.5 Neutro True Grit 1h 50m 10 7.6 Neutro American Hustle 2h 18m 10 7.2 Neutro Killers of the Flower Moon 3h 26m 10 7.6 Neutro The Banshees of Inisherin 1h 54m 9 7.7 Neutro The Sand Pebbles 3h 2m 8 7.5 Neutro The Elephant Man 2h 4m 8 8.2 Bom Ragtime 2h 35m 8 7.3 Neutro The Remains of the Day 2h 14m 8 7.8 Neutro Elvis 2h 39m 8 7.3 Neutro Pepe 3h 7 5.4 Ruim Mutiny on the Bounty 2h 58m 7 7.2 Neutro Hush...Hush, Sweet Charlotte 2h 13m 7 7.5 Neutro Hawaii 3h 9m 7 6.5 Ruim The Charge of the Light Brigade 2h 19m 7 6.6 Ruim Star! 2h 56m 7 6.4 Ruim Broadcast News 2h 13m 7 7.3 Neutro The Godfather Part III 2h 42m 7 7.6 Neutro The Prince of Tides 2h 12m 7 6.8 Ruim In the Name of the Father 2h 13m 7 8.1 Bom The Shawshank Redemption 2h 22m 7 9.3 Bom The Thin Red Line 2h 50m 7 7.6 Neutro The Insider 2h 37m 7 7.8 Neutro Seabiscuit 2h 20m 7 7.3 Neutro The Martian 2h 24m 7 8.0 Neutro The Fabelmans 2h 31m 7 7.5 Neutro Maestro 2h 9m 7 6.5 Ruim Billy Liar 1h 38m 6 7.2 Neutro The Cardinal 2h 55m 6 6.7 Ruim The Unsinkable Molly Brown 2h 8m 6 6.6 Ruim Lenny 1h 51m 6 7.5 Neutro The Knack... and How to Get It 1h 24m 6 6.3 Ruim Taking Off 1h 33m 6 7.4 Neutro Carol 1h 58m 6 7.3 Neutro Das Boot 2h 29m 6 8.4 Bom Empire of the Sun 2h 33m 6 7.7 Neutro Fatal Attraction 1h 59m 6 6.9 Ruim The Sixth Sense 1h 47m 6 8.2 Bom Good Night, and Good Luck. 1h 33m 6 7.4 Neutro Up in the Air 1h 49m 6 7.4 Neutro 127 Hours 1h 34m 6 7.5 Neutro Captain Phillips 2h 14m 6 7.8 Neutro Lion 1h 58m 6 8.0 Neutro The Trial of the Chicago 7 2h 9m 6 7.7 Neutro T�r 2h 38m 6 7.4 Neutro All of Us Strangers 1h 45m 6 7.6 Neutro Hook 2h 22m 5 6.8 Ruim Funny Lady 2h 16m 5 6.2 Ruim

Nesta última consulta, a título de curiosidade, acrescentei uma coluna que classifica o filme como Bom, Neutro ou Ruim segundo o critério descrito na introdução.  

![](https://storage.googleapis.com/mvps1_24/linha.jpg)

####Autoavaliação.
A proposta deste projeto foi mostrar de maneira símples, um pipeline de dados envolvendo a busca, coleta, modelagem, carga e análise dos dados utilizando tecnologias na nuvem. Foi feita a coleta e carga usando Pyspark e o Data Lake da Google Cloud Storage. Incluí também o processo de criação dos databases Bronze, Silver e Gold, onde o DB Bronze recebeu os dados crus, até o DB Gold onde foi finalizado o processo. Poderiam ter sido usadas outras library adicionais do Python, como o Pandas para formatar todas as tabelas, mas meu conhecimento sobre esta library ainda não me favorece na velocidade de produção, porém estou me atualizando sobre isto. 

Devido à minha total falta de experiência sobre esse universo de bancos de dados, estou conhecendo isto somente agora, precisei reestruturar todo o pipeline duas vezes. A primeira vez foi porque eu tinha feito uma coleta de um banco proveniente de uma base de dados da IMDb (ok), mas a segunda base de dados eu tinha armazenado no Databrick. Então achei que não seria o ideal e resolvi deslocar esta base de dados do Databricks para o Data Lake da Google Cloud Storage. A segunda vez que decidi refazer o pipeline foi porque notei que os nomes de várias tabelas não estavam padronizados, então refiz o procedimento todo para renomear as tabelas e os dataframes.

Com tudo, o objetivo foi alcançado. 
![](https://storage.googleapis.com/mvps1_24/linha.jpg)
 